components/user/Tabs/SettingsContent.tsx

```diff
@@ -1,59 +1,237 @@
+  graffiti: { ...FIELDS.graffiti },
+  country: { ...FIELDS.country, useDefault: false },
+  discord: {
+    ...FIELDS.social,
+    required: false,
+    options: undefined,
+    id: 'discord',
+    label: 'Discord',
+    placeholder: 'Your Discord username',
+    isRadioed: false,
+    validation: () => true,
+  },
+  telegram: {
+    ...FIELDS.social,
+    required: false,
+    options: undefined,
+    id: 'telegram',
+    label: 'Telegram',
+    placeholder: 'Your Telegram username',
+    isRadioed: false,
+    validation: () => true,
+  },
 }
 
-export default function SettingsContent({ authedUser }: Props) {
-  const $email = useField(FIELDS.email)
-  const $graffiti = useField(FIELDS.graffiti)
-  const $social = useField(FIELDS.social)
-  const $country = useField(FIELDS.country)
+export default function SettingsContent({
+  anyBlocksMined,
+  user,
+  authedUser,
+  toast,
+  reloadUser,
+  onTabChange,
+}: Props) {
+  const [$error, $setError] = useState<string>(UNSET)
+  const [$userData, $setUserData] = useState<API.ApiUserMetadata | null>(
+    authedUser
+  )
+
+  const {
+    email: _email = UNSET,
+    graffiti: _graffiti = UNSET,
+    discord: _discord = UNSET,
+    telegram: _telegram = UNSET,
+    country_code: _cc = UNSET,
+  } = $userData || {}
+  const $graffiti = useField({
+    ...EDITABLE_FIELDS.graffiti,
+    defaultValue: _graffiti,
+  })
+
+  const $email = useField({
+    ...EDITABLE_FIELDS.email,
+    defaultValue: _email,
+  })
+  const $discord = useField({
+    ...EDITABLE_FIELDS.discord,
+    defaultValue: _discord,
+    touched: !!_discord,
+  })
+  const $telegram = useField({
+    ...EDITABLE_FIELDS.telegram,
+    defaultValue: _telegram,
+    touched: !!_telegram,
+  })
 
-  const { choice: socialChoice, value: socialValue } =
-    getSocialChoice(authedUser)
+  const $country = useField({
+    ...EDITABLE_FIELDS.country,
+    defaultValue: _cc,
+  })
+  const testInvalid = useCallback(() => {
+    const invalid =
+      !$email?.valid ||
+      !$graffiti?.valid ||
+      !$discord?.valid ||
+      !$telegram?.valid ||
+      !$country?.valid
+    if (invalid) {
+      $setError('Please correct the invalid fields below')
+      scrollUp()
+    } else {
+      $setError(UNSET)
+    }
+    return invalid
+  }, [$email, $graffiti, $telegram, $discord, $country])
+  const update = useCallback(async () => {
+    if (
+      !$email ||
+      !$graffiti ||
+      !$telegram ||
+      !$discord ||
+      !$country ||
+      !authedUser ||
+      testInvalid()
+    ) {
+      return
+    }
+    const email = $email?.value
+    const graffiti = $graffiti?.value
+    const telegram = $telegram?.value
+    const discord = $discord?.value
+    const country = $country?.value
 
+    const updates = {
+      email,
+      graffiti,
+      telegram,
+      discord,
+      country_code: country,
+    }
+
+    const result = await API.updateUser(authedUser.id, updates)
+
+    if ('error' in result) {
```

Comment:
This isn't catching the `"Current graffiti '' has already mined blocks"` case -- it shows "User settings updated" instead.

I think the root cause is our API not returning consistent types for error responses, but I imagine that'll be a fair bit of work to get in order, so it'll probably be easier to adjust the frontend to account for it for now.

Reply:
Re-raising this issue, it's still something we should address before merging this.